File: gmall-canal/src/main/scala/com/vanas/canal/CanalClient.scala
package com.vanas.canal

import java.net.InetSocketAddress
import java.util

import com.alibaba.fastjson.JSONObject
import com.alibaba.otter.canal.client.{CanalConnector, CanalConnectors}
import com.alibaba.otter.canal.protocol.CanalEntry.{EventType, RowChange}
import com.alibaba.otter.canal.protocol.{CanalEntry, Message}
import com.google.protobuf.ByteString
import com.vanas.common.Constant

import scala.collection.JavaConversions._
import scala.util.Random

/**
 * @author Vanas
 * @create 2020-07-17 11:17 上午 
 */
object CanalClient {


    def sendToKafka(rowDatas: util.List[CanalEntry.RowData], topic: String) = {
        for (rowData <- rowDatas) {
            //变化后的列
            //mysql中的一行，到kafka的时候是一条
            val obj = new JSONObject()
            val columnList: util.List[CanalEntry.Column] = rowData.getAfterColumnsList
            for (column <- columnList) {
                //id:100 total_amount:1000.2
                val key: String = column.getName
                val value: String = column.getValue
                obj.put(key, value)
            }
            // println(obj.toJSONString)
            //4.解析后的数据，组成json字符串，写入到kafka
            new Thread() {
                override def run(): Unit = {
                    Thread.sleep(new Random().nextInt(20 * 1000))
                    MykafkaUtil.sendToKafka(topic, obj.toJSONString)
                }
            }.start()
            //MykafkaUtil.sendToKafka(topic, obj.toJSONString)
        }
    }

    def handleData(rowDatas: util.List[CanalEntry.RowData], tableName: String, eventType: CanalEntry.EventType) = {
        if (tableName == "order_info" && eventType == EventType.INSERT && rowDatas != null && !rowDatas.isEmpty) {
            val topic = Constant.TOPIC_ORDER_INFO
            sendToKafka(rowDatas, topic)
        } else if (tableName == "order_detail" && eventType == EventType.INSERT && rowDatas != null && !rowDatas.isEmpty) {
            val topic = Constant.TOPIC_ORDER_DETAIL
            sendToKafka(rowDatas, topic)
        }
    }

    def main(args: Array[String]): Unit = {
        //连接cannal
        val address = new InetSocketAddress("hadoop102", 11111)
        val connector: CanalConnector = CanalConnectors.newSingleConnector(address, "example", "", "")
        connector.connect() //连接canal
        //2拉取数据
        //2.1订阅想读的数据
        connector.subscribe("gmall_realtime.*")
        //2.2拉取数据
        while (true) {
            //100表示最多拉取由于100条sql导致变化的数据
            //所有的数据封装到一个message中
            val message: Message = connector.get(100)
            //解析数据
            val entries: util.List[CanalEntry.Entry] = message.getEntries
            //import scala.collection.JavaConversions._
            if (entries != null && !entries.isEmpty) {
                for (entry <- entries) {
                    //entry的类型必须是ROWDATA
                    if (entry != null && entry.hasEntryType && entry.getEntryType == CanalEntry.EntryType.ROWDATA) {
                        val value: ByteString = entry.getStoreValue
                        val rowChange: RowChange = RowChange.parseFrom(value)
                        //所有行变化的数据
                        val rowDatas: util.List[CanalEntry.RowData] = rowChange.getRowDatasList

                        handleData(rowDatas, entry.getHeader.getTableName, rowChange.getEventType)
                    }

                }
            } else { //没有拉取到数据
                println("没有拉取到数据，3秒后继续拉取")
                Thread.sleep(3000) //休眠3秒后继续拉取数据

            }
        }

        //把解析后的数据，组成json字符串，写入到kafka
    }
}

File: gmall-canal/src/main/java/com/vanas/client/CanalClient.java
package com.vanas.client;

import com.alibaba.fastjson.JSONObject;
import com.alibaba.otter.canal.client.CanalConnector;
import com.alibaba.otter.canal.client.CanalConnectors;
import com.alibaba.otter.canal.protocol.CanalEntry;
import com.alibaba.otter.canal.protocol.Message;
import com.google.protobuf.InvalidProtocolBufferException;

import java.net.InetSocketAddress;
import java.util.List;

/**
 * @author Vanas
 * @create 2020-07-16 9:07 下午
 */
public class CanalClient {
    public static void main(String[] args) {
        //获取Canal连接
        CanalConnector canalConnector = CanalConnectors.newSingleConnector(new InetSocketAddress("hadoop102", 11111), "example", "", "");

        //抓取数据
        while (true) {
            //连接canal
            canalConnector.connect();
            //指定订阅的数据库(gmall)
            canalConnector.subscribe("gamll_realtime.*");
            //执行抓取数据操作
            Message message = canalConnector.get(1024);
            //判断当前抓取的数据是否为空，如果为空，则休息一会
            if (message.getEntries().size() == 0) {
                System.out.println("没有数据，休息一会！");
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            } else {
                //解析message
                for (CanalEntry.Entry entry : message.getEntries()) {
                    //对当前entry类型做判断，只需要数据变化的内容,事务的开启与关闭以及心跳信息则过滤掉
                    if (CanalEntry.EntryType.ROWDATA.equals(entry.getEntryType())) {
                        //获取当前数据的表名
                        String tableName = entry.getHeader().getTableName();
                        try {
                            //将stroeValue反序列化
                            CanalEntry.RowChange rowChange = CanalEntry.RowChange.parseFrom(entry.getStoreValue());
                            //取出行集以及事件类型
                            List<CanalEntry.RowData> rowDatasList = rowChange.getRowDatasList();
                            CanalEntry.EventType eventType = rowChange.getEventType();
                            //处理数据
                            handler(tableName, rowDatasList, eventType);
                        } catch (InvalidProtocolBufferException e) {
                            e.printStackTrace();
                        }
                    }
                }

            }
        }
    }

    //处理监控的Mysql数据，解析并发送至Kafka
    private static void handler(String tableName, List<CanalEntry.RowData> rowDatasList, CanalEntry.EventType eventType) {
        //判断是否为订单表
        if (("order_info".equals(tableName))) {
            //取下单数据，即只要新增的数据
            if (CanalEntry.EventType.INSERT.equals(eventType)) {
                //遍历rowDataList
                for (CanalEntry.RowData rowData : rowDatasList) {
                    //创建JSON对象用于存放多个列的数据
                    JSONObject jsonObject = new JSONObject();
                    //遍历修改之后的数据列
                    for (CanalEntry.Column column : rowData.getAfterColumnsList()) {
                        jsonObject.put(column.getName(), column.getValue());
                    }
                    //打印当前行的数据,写入Kafka
                    System.out.println(jsonObject.toString());
                }
            }
        }
    }
}
